packages/react-compose/src/compose.ts

```diff
@@ -1,67 +1,29 @@
+import { mergeComposeOptions, wasComposedPreviously } from './utils';
 
-  mapPropsToStylesPropsChain: [],
+function compose<T extends React.ElementType, InputProps, InputStylesProps, ParentProps, ParentStylesProps>(
+  input: Input<T, InputProps>,
+  inputOptions: ComposeOptions<InputProps, InputStylesProps, ParentStylesProps> = {},
+) {
+  const composeOptions = mergeComposeOptions(
+    input as Input,
+    (inputOptions as unknown) as ComposeOptions,
+    wasComposedPreviously(input) ? input.fluentComposeConfig : undefined,
+  );
 
-  handledProps: [] as never[],
-  overrideStyles: false,
-};
+  const Component = (React.forwardRef<T, InputProps & ParentProps & { as?: React.ElementType }>((props, ref) => {
+    return composeOptions.render(props, ref as React.Ref<'div'>, composeOptions);
+  }) as unknown) as ComponentWithAs<T, InputProps & ParentProps>;
 
-function computeDisplayNames<InputProps, InputStylesProps, ParentProps, ParentStylesProps>(
-  inputOptions: ComposePreparedOptions,
-  composeOptions: ComposeOptions<InputProps, InputStylesProps, ParentStylesProps>,
-): string[] {
-  if (composeOptions.overrideStyles) {
-    return [composeOptions.displayName].filter(Boolean) as string[];
-  }
-
-  // To support styles composition we need to properly pick up display names
-  return composeOptions.displayName
-    ? inputOptions.displayNames.concat(composeOptions.displayName)
-    : inputOptions.displayNames;
-}
-
-function compose<InputProps, InputStylesProps, ParentProps, ParentStylesProps>(
-  InputComponent: React.FunctionComponent<ParentProps> & {
-    fluentComposeConfig?: ComposePreparedOptions;
-    handledProps?: (keyof ParentProps)[];
-  },
-  composeOptions: ComposeOptions<InputProps, InputStylesProps, ParentStylesProps> = {},
-): ComposedComponent<InputProps, InputStylesProps, ParentProps, ParentStylesProps> {
-  const Component = (InputComponent.bind(null) as unknown) as ComposedComponent<
-    InputProps,
-    InputStylesProps,
-    ParentProps,
-    ParentStylesProps
-  >;
+  Component.displayName = composeOptions.displayName;
 
-  const inputOptions: ComposePreparedOptions = InputComponent.fluentComposeConfig || {
-    ...defaultComposeOptions,
-    ...(InputComponent.displayName && { displayNames: [InputComponent.displayName] }),
-  };
-
-  Component.deprecated_className = composeOptions.className || inputOptions.className;
-  Component.displayName = composeOptions.displayName || InputComponent.displayName;
-
-  Component.fluentComposeConfig = {
-    className: Component.deprecated_className,
-    displayNames: computeDisplayNames(inputOptions, composeOptions),
-
-    mapPropsToStylesPropsChain: (composeOptions.mapPropsToStylesProps
-      ? [...inputOptions.mapPropsToStylesPropsChain, composeOptions.mapPropsToStylesProps]
-      : inputOptions.mapPropsToStylesPropsChain) as ((props: ParentStylesProps & InputProps) => InputStylesProps)[],
+  if ((input as InputComposeComponent).defaultProps) {
```

Comment:
@layershifter I guess this cast makes sense, as defaultProps will exists only on the components. Am I right?

Reply:
Yep